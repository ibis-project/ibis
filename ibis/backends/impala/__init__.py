"""Impala backend"""
import ibis.config
from ibis.backends.base import BaseBackend

# these objects are exposed in the public API and are not used in the module
from .client import (  # noqa: F401
    ImpalaClient,
    ImpalaConnection,
    ImpalaDatabase,
    ImpalaDatabaseTable,
    ImpalaTable,
)
from .hdfs import HDFS, WebHDFS, hdfs_connect
from .udf import *  # noqa: F401,F403


class Backend(BaseBackend):
    name = 'impala'
    kind = 'sql'
    client = ImpalaClient
    database_class = ImpalaDatabase
    table_class = ImpalaDatabaseTable
    table_expr_class = ImpalaTable
    HDFS = HDFS
    WebHDFS = WebHDFS

    def hdfs_connect(self, *args, **kwargs):
        return hdfs_connect(*args, **kwargs)

    def connect(
        self,
        host='localhost',
        port=21050,
        database='default',
        timeout=45,
        use_ssl=False,
        ca_cert=None,
        user=None,
        password=None,
        auth_mechanism='NOSASL',
        kerberos_service_name='impala',
        pool_size=8,
        hdfs_client=None,
    ):
        """Create an ImpalaClient for use with Ibis.

        Parameters
        ----------
        host : str, optional
            Host name of the impalad or HiveServer2 in Hive
        port : int, optional
            Impala's HiveServer2 port
        database : str, optional
            Default database when obtaining new cursors
        timeout : int, optional
            Connection timeout in seconds when communicating with HiveServer2
        use_ssl : bool, optional
            Use SSL when connecting to HiveServer2
        ca_cert : str, optional
            Local path to 3rd party CA certificate or copy of server
            certificate for self-signed certificates. If SSL is enabled, but
            this argument is ``None``, then certificate validation is skipped.
        user : str, optional
            LDAP user to authenticate
        password : str, optional
            LDAP password to authenticate
        auth_mechanism : str, optional
            {'NOSASL' <- default, 'PLAIN', 'GSSAPI', 'LDAP'}.
            Use NOSASL for non-secured Impala connections.  Use PLAIN for
            non-secured Hive clusters.  Use LDAP for LDAP authenticated
            connections.  Use GSSAPI for Kerberos-secured clusters.
        kerberos_service_name : str, optional
            Specify particular impalad service principal.

        Examples
        --------
        >>> import ibis
        >>> import os
        >>> hdfs_host = os.environ.get('IBIS_TEST_NN_HOST', 'localhost')
        >>> hdfs_port = int(os.environ.get('IBIS_TEST_NN_PORT', 50070))
        >>> impala_host = os.environ.get('IBIS_TEST_IMPALA_HOST', 'localhost')
        >>> impala_port = int(os.environ.get('IBIS_TEST_IMPALA_PORT', 21050))
        >>> hdfs = ibis.impala.hdfs_connect(host=hdfs_host, port=hdfs_port)
        >>> hdfs  # doctest: +ELLIPSIS
        <ibis.filesystems.WebHDFS object at 0x...>
        >>> client = ibis.impala.connect(
        ...     host=impala_host,
        ...     port=impala_port,
        ...     hdfs_client=hdfs,
        ... )
        >>> client  # doctest: +ELLIPSIS
        <ibis.impala.client.ImpalaClient object at 0x...>

        Returns
        -------
        ImpalaClient
        """
        params = {
            'host': host,
            'port': port,
            'database': database,
            'timeout': timeout,
            'use_ssl': use_ssl,
            'ca_cert': ca_cert,
            'user': user,
            'password': password,
            'auth_mechanism': auth_mechanism,
            'kerberos_service_name': kerberos_service_name,
        }

        con = ImpalaConnection(pool_size=pool_size, **params)
        try:
            client = ImpalaClient(
                backend=self, con=con, hdfs_client=hdfs_client
            )
        except Exception:
            con.close()
            raise
        return client

    def register_options(self):
        ibis.config.register_option(
            'temp_db',
            '__ibis_tmp',
            'Database to use for temporary tables, views. functions, etc.',
        )
        ibis.config.register_option(
            'temp_hdfs_path',
            '/tmp/ibis',
            'HDFS path for storage of temporary data',
        )
