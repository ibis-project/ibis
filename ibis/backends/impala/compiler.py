from __future__ import annotations

import ibis.expr.operations as ops
from ibis.backends.base.sql.compiler import Compiler, ExprTranslator, TableSetFormatter
from ibis.backends.base.sql.registry import binary_infix_ops, operation_registry, unary


class ImpalaTableSetFormatter(TableSetFormatter):
    def _get_join_type(self, op):
        jname = self._join_names[type(op)]

        # Impala requires this
        if not op.predicates:
            jname = self._join_names[ops.CrossJoin]

        return jname

    def _format_in_memory_table(self, op):
        if op.data:
            return super()._format_in_memory_table(op)

        schema = op.schema
        names = schema.names
        types = schema.types
        rows = [
            f"{self._translate(ops.Cast(ops.Literal(None, dtype=dtype), to=dtype))} AS {name}"
            for name, dtype in zip(map(self._quote_identifier, names), types)
        ]
        return f"(SELECT * FROM (SELECT {', '.join(rows)}) AS _ LIMIT 0)"


class ImpalaExprTranslator(ExprTranslator):
    _registry = {**operation_registry, **binary_infix_ops, ops.Hash: unary("fnv_hash")}
    _forbids_frame_clause = (
        *ExprTranslator._forbids_frame_clause,
        ops.Lag,
        ops.Lead,
        ops.FirstValue,
        ops.LastValue,
    )
    _unsupported_reductions = (
        ops.ApproxMedian,
        ops.ApproxCountDistinct,
        ops.GroupConcat,
    )
    _dialect_name = "hive"
    _quote_identifiers = True


rewrites = ImpalaExprTranslator.rewrites


@rewrites(ops.FloorDivide)
def _floor_divide(op):
    return ops.Floor(ops.Divide(op.left, op.right))


class ImpalaCompiler(Compiler):
    translator_class = ImpalaExprTranslator
    table_set_formatter_class = ImpalaTableSetFormatter
