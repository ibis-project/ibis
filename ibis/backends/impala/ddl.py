from __future__ import annotations

# Copyright 2014 Cloudera Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
import json

from ibis.backends.base.sql.ddl import (
    AlterTable,
    BaseDDL,
    CreateTable,
    CreateTableWithSchema,
    DropFunction,
    format_partition,
    format_schema,
    format_tblproperties,
)
from ibis.backends.base.sql.registry import type_to_sql_string


class CreateTableParquet(CreateTable):
    def __init__(
        self,
        table_name,
        path,
        example_file=None,
        example_table=None,
        schema=None,
        external=True,
        **kwargs,
    ):
        super().__init__(
            table_name,
            external=external,
            format="parquet",
            path=path,
            **kwargs,
        )
        self.example_file = example_file
        self.example_table = example_table
        self.schema = schema

    @property
    def _pieces(self):
        if self.example_file is not None:
            yield f"LIKE PARQUET '{self.example_file}'"
        elif self.example_table is not None:
            yield f"LIKE {self.example_table}"
        elif self.schema is not None:
            yield format_schema(self.schema)
        else:
            raise NotImplementedError

        yield self._storage()
        yield self._location()


class DelimitedFormat:
    def __init__(
        self,
        path,
        delimiter=None,
        escapechar=None,
        na_rep=None,
        lineterminator=None,
    ):
        self.path = path
        self.delimiter = delimiter
        self.escapechar = escapechar
        self.lineterminator = lineterminator
        self.na_rep = na_rep

    def to_ddl(self):
        yield "ROW FORMAT DELIMITED"

        if self.delimiter is not None:
            yield f"FIELDS TERMINATED BY '{self.delimiter}'"

        if self.escapechar is not None:
            yield f"ESCAPED BY '{self.escapechar}'"

        if self.lineterminator is not None:
            yield f"LINES TERMINATED BY '{self.lineterminator}'"

        yield "STORED AS TEXTFILE"
        yield f"LOCATION '{self.path}'"

        if self.na_rep is not None:
            props = {"serialization.null.format": self.na_rep}
            yield format_tblproperties(props)


class AvroFormat:
    def __init__(self, path, avro_schema):
        self.path = path
        self.avro_schema = avro_schema

    def to_ddl(self):
        yield "STORED AS AVRO"
        yield f"LOCATION '{self.path}'"

        schema = json.dumps(self.avro_schema, indent=2, sort_keys=True)
        schema = "\n".join(x.rstrip() for x in schema.splitlines())

        props = {"avro.schema.literal": schema}
        yield format_tblproperties(props)


class ParquetFormat:
    def __init__(self, path):
        self.path = path

    def to_ddl(self):
        yield "STORED AS PARQUET"
        yield f"LOCATION '{self.path}'"


class CreateTableDelimited(CreateTableWithSchema):
    def __init__(
        self,
        table_name,
        path,
        schema,
        delimiter=None,
        escapechar=None,
        lineterminator=None,
        na_rep=None,
        external=True,
        **kwargs,
    ):
        table_format = DelimitedFormat(
            path,
            delimiter=delimiter,
            escapechar=escapechar,
            lineterminator=lineterminator,
            na_rep=na_rep,
        )
        super().__init__(table_name, schema, table_format, external=external, **kwargs)


class CreateTableAvro(CreateTable):
    def __init__(self, table_name, path, avro_schema, external=True, **kwargs):
        super().__init__(table_name, external=external, **kwargs)
        self.table_format = AvroFormat(path, avro_schema)

    @property
    def _pieces(self):
        yield "\n".join(self.table_format.to_ddl())


class LoadData(BaseDDL):
    """Generate DDL for LOAD DATA command.

    Cannot be cancelled
    """

    def __init__(
        self,
        table_name,
        path,
        database=None,
        partition=None,
        partition_schema=None,
        overwrite=False,
    ):
        self.table_name = table_name
        self.database = database
        self.path = path

        self.partition = partition
        self.partition_schema = partition_schema

        self.overwrite = overwrite

    def compile(self):
        overwrite = "OVERWRITE " if self.overwrite else ""

        if self.partition is not None:
            partition = "\n" + format_partition(self.partition, self.partition_schema)
        else:
            partition = ""

        scoped_name = self._get_scoped_name(self.table_name, self.database)
        return "LOAD DATA INPATH '{}' {}INTO TABLE {}{}".format(
            self.path, overwrite, scoped_name, partition
        )


class PartitionProperties(AlterTable):
    def __init__(
        self,
        table,
        partition,
        partition_schema,
        location=None,
        format=None,
        tbl_properties=None,
        serde_properties=None,
    ):
        super().__init__(
            table,
            location=location,
            format=format,
            tbl_properties=tbl_properties,
            serde_properties=serde_properties,
        )
        self.partition = partition
        self.partition_schema = partition_schema

    def _compile(self, cmd, property_prefix=""):
        part = format_partition(self.partition, self.partition_schema)
        if cmd:
            part = f"{cmd} {part}"

        props = self._format_properties(property_prefix)
        action = f"{self.table} {part}{props}"
        return self._wrap_command(action)


class AddPartition(PartitionProperties):
    def __init__(self, table, partition, partition_schema, location=None):
        super().__init__(table, partition, partition_schema, location=location)

    def compile(self):
        return self._compile("ADD")


class AlterPartition(PartitionProperties):
    def compile(self):
        return self._compile("", "SET ")


class DropPartition(PartitionProperties):
    def __init__(self, table, partition, partition_schema):
        super().__init__(table, partition, partition_schema)

    def compile(self):
        return self._compile("DROP")


class CacheTable(BaseDDL):
    def __init__(self, table_name, database=None, pool="default"):
        self.table_name = table_name
        self.database = database
        self.pool = pool

    def compile(self):
        scoped_name = self._get_scoped_name(self.table_name, self.database)
        return f"ALTER TABLE {scoped_name} SET CACHED IN '{self.pool}'"


class CreateFunction(BaseDDL):
    _object_type = "FUNCTION"

    def __init__(self, func, name=None, database=None):
        self.func = func
        self.name = name or func.name
        self.database = database

    def _impala_signature(self):
        scoped_name = self._get_scoped_name(self.name, self.database)
        input_sig = _impala_input_signature(self.func.inputs)
        output_sig = type_to_sql_string(self.func.output)

        return f"{scoped_name}({input_sig}) returns {output_sig}"


class CreateUDF(CreateFunction):
    def compile(self):
        create_decl = "CREATE FUNCTION"
        impala_sig = self._impala_signature()
        param_line = f"location '{self.func.lib_path}' symbol='{self.func.so_symbol}'"
        return f"{create_decl} {impala_sig} {param_line}"


class CreateUDA(CreateFunction):
    def compile(self):
        create_decl = "CREATE AGGREGATE FUNCTION"
        impala_sig = self._impala_signature()
        tokens = [f"location '{self.func.lib_path}'"]

        fn_names = (
            "init_fn",
            "update_fn",
            "merge_fn",
            "serialize_fn",
            "finalize_fn",
        )

        for fn in fn_names:
            value = getattr(self.func, fn)
            if value is not None:
                tokens.append(f"{fn}='{value}'")

        joined_tokens = "\n".join(tokens)
        return f"{create_decl} {impala_sig} {joined_tokens}"


class DropFunction(DropFunction):
    def _impala_signature(self):
        full_name = self._get_scoped_name(self.name, self.database)
        input_sig = _impala_input_signature(self.inputs)
        return f"{full_name}({input_sig})"


class ListFunction(BaseDDL):
    def __init__(self, database, like=None, aggregate=False):
        self.database = database
        self.like = like
        self.aggregate = aggregate

    def compile(self):
        statement = "SHOW "
        if self.aggregate:
            statement += "AGGREGATE "
        statement += f"FUNCTIONS IN {self.database}"
        if self.like:
            statement += f" LIKE '{self.like}'"
        return statement


def _impala_input_signature(inputs):
    # TODO: varargs '{}...'.format(val)
    return ", ".join(map(type_to_sql_string, inputs))
