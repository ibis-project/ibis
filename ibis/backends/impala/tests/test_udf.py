from __future__ import annotations

from decimal import Decimal
from posixpath import join as pjoin

import numpy as np
import pandas as pd
import pytest

import ibis
import ibis.backends.impala as api
import ibis.common.exceptions as com
import ibis.expr.datatypes as dt
import ibis.expr.types as ir
from ibis import util
from ibis.backends.impala import ddl
from ibis.common.annotations import ValidationError
from ibis.expr import rules

pytest.importorskip("impala")


@pytest.fixture(scope="module")
def table(mockcon):
    return mockcon.table("functional_alltypes")


@pytest.fixture
def i8(table):
    return table.tinyint_col


@pytest.fixture
def i16(table):
    return table.smallint_col


@pytest.fixture
def i32(table):
    return table.int_col


@pytest.fixture
def i64(table):
    return table.bigint_col


@pytest.fixture
def d(table):
    return table.double_col


@pytest.fixture
def f(table):
    return table.float_col


@pytest.fixture
def s(table):
    return table.string_col


@pytest.fixture
def b(table):
    return table.bool_col


@pytest.fixture
def t(table):
    return table.timestamp_col


@pytest.fixture
def tpch_customer(con):
    return con.table("customer")


@pytest.fixture
def dec(tpch_customer):
    return tpch_customer.c_acctbal


@pytest.fixture
def all_cols(i8, i16, i32, i64, d, f, dec, s, b, t):
    return [
        i8,
        i16,
        i32,
        i64,
        d,
        f,
        dec,
        s,
        b,
        t,
    ]


def test_sql_generation(snapshot):
    func = api.scalar_function(["string"], "string", name="Tester")
    func.register("identity", "udf_testing")

    result = func("hello world")
    snapshot.assert_match(ibis.impala.compile(result), "out.sql")


def test_sql_generation_from_infoclass(snapshot):
    func = api.wrap_udf("test.so", ["string"], "string", "info_test")
    repr(func)

    func.register("info_test", "udf_testing")
    result = func("hello world").name("tmp")
    snapshot.assert_match(ibis.impala.compile(result), "out.sql")


@pytest.mark.parametrize(
    ("ty", "value", "column"),
    [
        pytest.param("boolean", True, "bool_col", id="boolean"),
        pytest.param("int8", 1, "tinyint_col", id="int8"),
        pytest.param("int16", 1, "smallint_col", id="int16"),
        pytest.param("int32", 1, "int_col", id="int32"),
        pytest.param("int64", 1, "bigint_col", id="int64"),
        pytest.param("float", 1.0, "float_col", id="float"),
        pytest.param("double", 1.0, "double_col", id="double"),
        pytest.param("string", "1", "string_col", id="string"),
        pytest.param(
            "timestamp",
            ibis.timestamp("1961-04-10"),
            "timestamp_col",
            id="timestamp",
        ),
    ],
)
def test_udf_primitive_output_types(ty, value, column, table):
    func = _register_udf([ty], ty, "test")

    ibis_type = dt.validate_type(ty)

    expr = func(value)
    assert type(expr) == getattr(ir, ibis_type.scalar)

    expr = func(table[column])
    assert type(expr) == getattr(ir, ibis_type.column)


@pytest.mark.parametrize(
    ("ty", "value"),
    [
        pytest.param("boolean", True, id="boolean"),
        pytest.param("int8", 1, id="int8"),
        pytest.param("int16", 1, id="int16"),
        pytest.param("int32", 1, id="int32"),
        pytest.param("int64", 1, id="int64"),
        pytest.param("float", 1.0, id="float"),
        pytest.param("double", 1.0, id="double"),
        pytest.param("string", "1", id="string"),
        pytest.param(
            "timestamp",
            ibis.timestamp("1961-04-10"),
            id="timestamp",
        ),
    ],
)
def test_uda_primitive_output_types(ty, value):
    func = _register_uda([ty], ty, "test")

    ibis_type = dt.validate_type(ty)
    scalar_type = getattr(ir, ibis_type.scalar)

    expr1 = func(value)
    assert isinstance(expr1, scalar_type)

    expr2 = func(value)
    assert isinstance(expr2, scalar_type)


def test_decimal(dec):
    func = _register_udf(["decimal(12, 2)"], "decimal(12, 2)", "test")
    expr = func(1.0)
    assert type(expr) == ir.DecimalScalar
    expr = func(dec)
    assert type(expr) == ir.DecimalColumn


@pytest.mark.parametrize(
    ("ty", "valid_cast_indexer"),
    [
        pytest.param("decimal(12, 2)", slice(7), id="decimal"),
        pytest.param("double", slice(6), id="double"),
        pytest.param("float", slice(6), id="float"),
        pytest.param("int16", slice(2), id="int16"),
        pytest.param("int32", slice(3), id="int32"),
        pytest.param("int64", slice(4), id="int64"),
        pytest.param("int8", slice(1), id="int8"),
    ],
)
def test_udf_valid_typecasting(ty, valid_cast_indexer, all_cols):
    func = _register_udf([ty], "int32", "typecast")

    for expr in all_cols[valid_cast_indexer]:
        func(expr)


@pytest.mark.parametrize(
    ("ty", "valid_cast_indexer"),
    [
        pytest.param("boolean", slice(8), id="boolean_first_8"),
        pytest.param("boolean", slice(9, None), id="boolean_9_onwards"),
        pytest.param("decimal", slice(7, None), id="decimal"),
        pytest.param("double", slice(-3, None), id="double"),
        pytest.param("float", slice(-3, None), id="float"),
        pytest.param("int16", slice(2, None), id="int16"),
        pytest.param("int32", slice(3, None), id="int32"),
        pytest.param("int64", slice(4, None), id="int64"),
        pytest.param("int8", slice(1, None), id="int8"),
        pytest.param("string", slice(7), id="string_first_7"),
        pytest.param("string", slice(8, None), id="string_8_onwards"),
        pytest.param("timestamp", slice(-1), id="timestamp"),
    ],
)
def test_udf_invalid_typecasting(ty, valid_cast_indexer, all_cols):
    func = _register_udf([ty], "int32", "typecast")

    for expr in all_cols[valid_cast_indexer]:
        with pytest.raises(ValidationError):
            func(expr)


def test_mult_args(i32, d, s, b, t):
    func = _register_udf(
        ["int32", "double", "string", "boolean", "timestamp"],
        "int64",
        "mult_types",
    )

    expr = func(i32, d, s, b, t)
    assert issubclass(type(expr), ir.Column)

    expr = func(1, 1.0, "a", True, ibis.timestamp("1961-04-10"))
    assert issubclass(type(expr), ir.Scalar)


def _register_udf(inputs, output, name):
    func = api.scalar_function(inputs, output, name=name)
    func.register(name, "ibis_testing")
    return func


def _register_uda(inputs, output, name):
    func = api.aggregate_function(inputs, output, name=name)
    func.register(name, "ibis_testing")
    return func


@pytest.fixture
def udfcon(con, monkeypatch):
    monkeypatch.setitem(con.con.options, "DISABLE_CODEGEN", "0")
    return con


@pytest.fixture
def alltypes(udfcon):
    return udfcon.table("functional_alltypes")


@pytest.fixture
def udf_ll(test_data_dir):
    return pjoin(test_data_dir, "udf/udf-sample.ll")


@pytest.fixture
def uda_ll(test_data_dir):
    return pjoin(test_data_dir, "udf/uda-sample.ll")


@pytest.fixture
def uda_so(test_data_dir):
    return pjoin(test_data_dir, "udf/libudasample.so")


@pytest.mark.parametrize(
    ("typ", "lit_val", "col_name"),
    [
        pytest.param("boolean", True, "bool_col", id="boolean"),
        pytest.param("int8", ibis.literal(5), "tinyint_col", id="int8"),
        pytest.param(
            "int16",
            ibis.literal(2**10),
            "smallint_col",
            id="int16",
        ),
        pytest.param("int32", ibis.literal(2**17), "int_col", id="int16"),
        pytest.param("int64", ibis.literal(2**33), "bigint_col", id="int64"),
        pytest.param("float", ibis.literal(3.14), "float_col", id="float"),
        pytest.param("double", ibis.literal(3.14), "double_col", id="double"),
        pytest.param(
            "string",
            ibis.literal("ibis"),
            "string_col",
            id="string",
        ),
        pytest.param(
            "timestamp",
            ibis.timestamp("1961-04-10"),
            "timestamp_col",
            id="timestamp",
        ),
    ],
)
@pytest.mark.xfail(
    reason="Unknown reason. xfailing to restore the CI for udf tests. #2358"
)
def test_identity_primitive_types(
    udfcon, alltypes, test_data_db, udf_ll, typ, lit_val, col_name
):
    col_val = alltypes[col_name]
    identity_func_testing(udf_ll, udfcon, test_data_db, typ, lit_val, col_val)


@pytest.mark.xfail(
    reason="Unknown reason. xfailing to restore the CI for udf tests. #2358"
)
def test_decimal_fail(udfcon, test_data_db, udf_ll):
    col = udfcon.table("customer").c_acctbal
    literal = ibis.literal(1).cast("decimal(12,2)")
    name = "__tmp_udf_" + util.guid()

    func = udf_creation_to_op(
        udf_ll,
        udfcon,
        test_data_db,
        name,
        "Identity",
        ["decimal(12,2)"],
        "decimal(12,2)",
    )

    expr = func(literal)
    assert issubclass(type(expr), ir.Scalar)
    result = udfcon.execute(expr)
    assert result == Decimal(1)

    expr = func(col)
    assert issubclass(type(expr), ir.Column)
    udfcon.execute(expr)


@pytest.mark.xfail(
    reason="Unknown reason. xfailing to restore the CI for udf tests. #2358"
)
def test_mixed_inputs(udfcon, alltypes, test_data_db, udf_ll):
    name = "two_args"
    symbol = "TwoArgs"
    inputs = ["int32", "int32"]
    output = "int32"
    func = udf_creation_to_op(
        udf_ll, udfcon, test_data_db, name, symbol, inputs, output
    )

    expr = func(alltypes.int_col, 1)
    assert issubclass(type(expr), ir.Column)
    udfcon.execute(expr)

    expr = func(1, alltypes.int_col)
    assert issubclass(type(expr), ir.Column)
    udfcon.execute(expr)

    expr = func(alltypes.int_col, alltypes.tinyint_col)
    udfcon.execute(expr)


@pytest.mark.xfail(
    reason="Unknown reason. xfailing to restore the CI for udf tests. #2358"
)
def test_implicit_typecasting(udfcon, alltypes, test_data_db, udf_ll):
    col = alltypes.tinyint_col
    literal = ibis.literal(1000)
    identity_func_testing(udf_ll, udfcon, test_data_db, "int32", literal, col)


def identity_func_testing(udf_ll, udfcon, test_data_db, datatype, literal, column):
    inputs = [datatype]
    name = "__tmp_udf_" + util.guid()
    func = udf_creation_to_op(
        udf_ll, udfcon, test_data_db, name, "Identity", inputs, datatype
    )

    expr = func(literal)
    assert issubclass(type(expr), ir.Scalar)
    result = udfcon.execute(expr)
    # Hacky
    if datatype == "timestamp":
        assert type(result) == pd.Timestamp
    else:
        lop = literal.op()
        if isinstance(lop, ir.Literal):
            np.testing.assert_allclose(lop.value, 5)
        else:
            np.testing.assert_allclose(result, udfcon.execute(literal), 5)

    expr = func(column)
    assert issubclass(type(expr), ir.Column)
    udfcon.execute(expr)


@pytest.mark.xfail(
    reason="Unknown reason. xfailing to restore the CI for udf tests. #2358"
)
def test_mult_type_args(udfcon, alltypes, test_data_db, udf_ll):
    symbol = "AlmostAllTypes"
    name = "most_types"
    inputs = [
        "string",
        "boolean",
        "int8",
        "int16",
        "int32",
        "int64",
        "float",
        "double",
    ]
    output = "int32"

    func = udf_creation_to_op(
        udf_ll, udfcon, test_data_db, name, symbol, inputs, output
    )

    expr = func("a", True, 1, 1, 1, 1, 1.0, 1.0)
    result = udfcon.execute(expr)
    assert result == 8

    table = alltypes
    expr = func(
        table.string_col,
        table.bool_col,
        table.tinyint_col,
        table.tinyint_col,
        table.smallint_col,
        table.smallint_col,
        1.0,
        1.0,
    )
    udfcon.execute(expr)


@pytest.mark.xfail(
    reason="Unknown reason. xfailing to restore the CI for udf tests. #2358"
)
def test_udf_varargs(udfcon, alltypes, udf_ll, test_data_db):
    t = alltypes

    name = f"add_numbers_{util.guid()[:4]}"

    input_sig = rules.varargs(rules.double)
    func = api.wrap_udf(udf_ll, input_sig, "double", "AddNumbers", name=name)
    func.register(name, test_data_db)
    udfcon.create_function(func, database=test_data_db)

    expr = func(t.double_col, t.double_col)
    expr.execute()


def test_drop_udf_not_exists(udfcon):
    random_name = util.guid()
    with pytest.raises(com.MissingUDFError, match=random_name):
        udfcon.drop_udf(random_name)


def test_drop_uda_not_exists(udfcon):
    random_name = util.guid()
    with pytest.raises(com.MissingUDFError, match=random_name):
        udfcon.drop_uda(random_name)


def udf_creation_to_op(udf_ll, udfcon, test_data_db, name, symbol, inputs, output):
    func = api.wrap_udf(udf_ll, inputs, output, symbol, name)

    udfcon.create_function(func, database=test_data_db)

    func.register(name, test_data_db)

    assert udfcon.exists_udf(name, test_data_db)
    return func


def test_ll_uda_not_supported(uda_ll):
    # LLVM IR UDAs are not supported as of Impala 2.2
    with pytest.raises(com.IbisError):
        conforming_wrapper(uda_ll, ["double"], "double", "Variance")


def conforming_wrapper(where, inputs, output, prefix, serialize=True, name=None):
    kwds = {"name": name}
    if serialize:
        kwds["serialize_fn"] = f"{prefix}Serialize"
    return api.wrap_uda(
        where,
        inputs,
        output,
        f"{prefix}Update",
        init_fn=f"{prefix}Init",
        merge_fn=f"{prefix}Merge",
        finalize_fn=f"{prefix}Finalize",
        **kwds,
    )


@pytest.fixture
def wrapped_count_uda(uda_so):
    name = f"user_count_{util.guid()}"
    return api.wrap_uda(uda_so, ["int32"], "int64", "CountUpdate", name=name)


def test_count_uda(udfcon, alltypes, test_data_db, wrapped_count_uda):
    func = wrapped_count_uda
    func.register(func.name, test_data_db)
    udfcon.create_function(func, database=test_data_db)

    # it works!
    func(alltypes.int_col).execute()


def test_list_udas(udfcon, temp_database, wrapped_count_uda):
    func = wrapped_count_uda
    db = temp_database
    udfcon.create_function(func, database=db)

    funcs = udfcon.list_udas(database=db)

    f = funcs[0]
    assert f.name == func.name
    assert f.inputs == func.inputs
    assert f.output == func.output


@pytest.mark.xfail(
    reason="Unknown reason. xfailing to restore the CI for udf tests. #2358"
)
def test_drop_database_with_udfs_and_udas(udfcon, temp_database, wrapped_count_uda):
    uda1 = wrapped_count_uda

    udf1 = api.wrap_udf(
        udf_ll,
        ["boolean"],
        "boolean",
        "Identity",
        f"udf_{util.guid()}",
    )

    db = temp_database

    udfcon.create_database(db)

    udfcon.create_function(uda1, database=db)
    udfcon.create_function(udf1, database=db)
    # drop happens in test tear down


@pytest.fixture
def inputs():
    return ["string", "string"]


@pytest.fixture
def output():
    return "int64"


@pytest.fixture
def name():
    return "test_name"


def test_create_udf(inputs, output, name, snapshot):
    func = api.wrap_udf(
        "/foo/bar.so",
        inputs,
        output,
        so_symbol="testFunc",
        name=name,
    )
    stmt = ddl.CreateUDF(func)
    result = stmt.compile()
    snapshot.assert_match(result, "out.sql")


def test_create_udf_type_conversions(output, name, snapshot):
    inputs = ["string", "int8", "int16", "int32"]
    func = api.wrap_udf(
        "/foo/bar.so",
        inputs,
        output,
        so_symbol="testFunc",
        name=name,
    )
    stmt = ddl.CreateUDF(func)
    result = stmt.compile()
    snapshot.assert_match(result, "out.sql")


def test_delete_udf_simple(name, inputs, snapshot):
    stmt = ddl.DropFunction(name, inputs)
    result = stmt.compile()
    snapshot.assert_match(result, "out.sql")


def test_delete_udf_if_exists(name, inputs, snapshot):
    stmt = ddl.DropFunction(name, inputs, must_exist=False)
    result = stmt.compile()
    snapshot.assert_match(result, "out.sql")


def test_delete_udf_aggregate(name, inputs, snapshot):
    stmt = ddl.DropFunction(name, inputs, aggregate=True)
    result = stmt.compile()
    snapshot.assert_match(result, "out.sql")


def test_delete_udf_db(name, inputs, snapshot):
    stmt = ddl.DropFunction(name, inputs, database="test")
    result = stmt.compile()
    snapshot.assert_match(result, "out.sql")


@pytest.mark.parametrize("series", [True, False])
def test_create_uda(name, inputs, output, series, snapshot):
    func = api.wrap_uda(
        "/foo/bar.so",
        inputs,
        output,
        update_fn="Update",
        init_fn="Init",
        merge_fn="Merge",
        finalize_fn="Finalize",
        serialize_fn="Serialize" if series else None,
    )
    stmt = ddl.CreateUDA(func, name=name, database="bar")
    result = stmt.compile()
    snapshot.assert_match(result, "out.sql")


def test_list_udf(snapshot):
    stmt = ddl.ListFunction("test")
    result = stmt.compile()
    snapshot.assert_match(result, "out.sql")


def test_list_udfs_like(snapshot):
    stmt = ddl.ListFunction("test", like="identity")
    result = stmt.compile()
    snapshot.assert_match(result, "out.sql")


def test_list_udafs(snapshot):
    stmt = ddl.ListFunction("test", aggregate=True)
    result = stmt.compile()
    snapshot.assert_match(result, "out.sql")


def test_list_udafs_like(snapshot):
    stmt = ddl.ListFunction("test", like="identity", aggregate=True)
    result = stmt.compile()
    snapshot.assert_match(result, "out.sql")
