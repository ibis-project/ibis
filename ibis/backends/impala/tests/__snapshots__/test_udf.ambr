# name: test_create_uda[False]
  '''
  CREATE AGGREGATE FUNCTION bar.`test_name`(string, string) returns bigint location '/foo/bar.so'
  init_fn='Init'
  update_fn='Update'
  merge_fn='Merge'
  finalize_fn='Finalize'
  '''
# ---
# name: test_create_uda[True]
  '''
  CREATE AGGREGATE FUNCTION bar.`test_name`(string, string) returns bigint location '/foo/bar.so'
  init_fn='Init'
  update_fn='Update'
  merge_fn='Merge'
  serialize_fn='Serialize'
  finalize_fn='Finalize'
  '''
# ---
# name: test_create_udf
  "CREATE FUNCTION `test_name`(string, string) returns bigint location '/foo/bar.so' symbol='testFunc'"
# ---
# name: test_create_udf_type_conversions
  "CREATE FUNCTION `test_name`(string, tinyint, smallint, int) returns bigint location '/foo/bar.so' symbol='testFunc'"
# ---
# name: test_delete_udf_aggregate
  'DROP AGGREGATE FUNCTION `test_name`(string, string)'
# ---
# name: test_delete_udf_db
  'DROP FUNCTION test.`test_name`(string, string)'
# ---
# name: test_delete_udf_if_exists
  'DROP FUNCTION IF EXISTS `test_name`(string, string)'
# ---
# name: test_delete_udf_simple
  'DROP FUNCTION `test_name`(string, string)'
# ---
# name: test_list_udafs
  'SHOW AGGREGATE FUNCTIONS IN test'
# ---
# name: test_list_udafs_like
  "SHOW AGGREGATE FUNCTIONS IN test LIKE 'identity'"
# ---
# name: test_list_udf
  'SHOW FUNCTIONS IN test'
# ---
# name: test_list_udfs_like
  "SHOW FUNCTIONS IN test LIKE 'identity'"
# ---
# name: test_sql_generation
  "SELECT udf_testing.identity('hello world') AS `UDF_Tester('hello world')`"
# ---
# name: test_sql_generation_from_infoclass
  "SELECT udf_testing.info_test('hello world') AS `tmp`"
# ---
